File: src/Game.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Scanner;

class Move {
  private Integer r = 0;
  private Integer c = 0;

  public Move(Integer r, Integer c) {
    this.r = r;
    this.c = c;
  }

  public Integer getR() {
    return r;
  }

  public Integer getC() {
    return c;
  }
}

class Node {
  private Move m = null;
  private Integer value = 0;

  public Node(Move m, Integer value) {
    this.m = m;
    this.value = value;
  }

  public Move getMove() {
    return m;
  }

  public Integer getValue() {
    return value;
  }
}

class Board {
  private static int[][] cells = new int[3][3];
  private static List<Move> availableMoves = new ArrayList<>();
  private static List<Node> nodes = new ArrayList<>();
  private static Character[][] points = new Character[3][3];

  static void placeMove(Move m, Integer i) {
    cells[m.getR()][m.getC()] = i;
  }

  static List<Move> getAvailableMoves() {
    if (availableMoves.size() > 0) {
      availableMoves.clear();
    }
    for (int i = 0; i < 3; i++) {
      for (int j = 0; j < 3; j++) {
        if (cells[i][j] == 0) {
          availableMoves.add(new Move(i, j));
        }
      }
    }
    return availableMoves;
  }

  static Boolean xHasWon() {
    for (int i = 0; i < 3; i++) {
      if (cells[i][0] == cells[i][1] && cells[i][1] == cells[i][2] && cells[i][2] == 1) {
        return true;
      }
      if (cells[0][i] == cells[1][i] && cells[1][i] == cells[2][i] && cells[2][i] == 1) {
        return true;
      }
    }
    if ((cells[0][0] == cells[1][1] && cells[1][1] == cells[2][2] && cells[2][2] == 1) || (cells[0][2] == cells[1][1] && cells[1][1] == cells[2][0] && cells[2][0] == 1)) {
      return true;
    }
    return false;
  }

  static Boolean oHasWon() {
    for (int i = 0; i < 3; i++) {
      if (cells[i][0] == cells[i][1] && cells[i][1] == cells[i][2] && cells[i][2] == -1) {
        return true;
      }
      if (cells[0][i] == cells[1][i] && cells[1][i] == cells[2][i] && cells[2][i] == -1) {
        return true;
      }
    }
    if ((cells[0][0] == cells[1][1] && cells[1][1] == cells[2][2] && cells[2][2] == -1) || (cells[0][2] == cells[1][1] && cells[1][1] == cells[2][0] && cells[2][0] == -1)) {
      return true;
    }
    return false;
  }

  static Boolean gameIsOver() {
    return (xHasWon() || oHasWon() || getAvailableMoves().isEmpty());
  }

  static Integer minimax(Integer depth, Boolean isMaximizer) {
    Integer maxEval = Integer.MIN_VALUE;
    Integer minEval = Integer.MAX_VALUE;
    if (getAvailableMoves().isEmpty()) {
      return 0;
    }
    if (xHasWon()) {
      return 1;
    }
    if (oHasWon()) {
      return -1;
    }
    if (isMaximizer) {
      for (int i = 0; i < getAvailableMoves().size(); i++) {
        Move m = getAvailableMoves().get(i);
        placeMove(m, 1);
        Integer eval = minimax(depth + 1, false);
        maxEval = Math.max(maxEval, eval);
        if (depth == 0) {
          nodes.add(new Node(m, maxEval));
        }
        placeMove(m, 0);
      }
    } else {
      for (int i = 0; i < getAvailableMoves().size(); i++) {
        Move m = getAvailableMoves().get(i);
        placeMove(m, -1);
        Integer eval = minimax(depth + 1, true);
        minEval = Math.min(minEval, eval);
        if (depth == 0) {
          nodes.add(new Node(m, minEval));
        }
        placeMove(m, 0);
      }
    }
    return isMaximizer ? maxEval : minEval;
  }

  static Move getBestMove() {
    Integer max = Integer.MIN_VALUE;
    Integer index = 0;
    for (int i = 0; i < nodes.size(); i++) {
      Node n = nodes.get(i);
      if (max < n.getValue()) {
        max = n.getValue();
        index = i;
      }
    }
    return nodes.get(index).getMove();
  }

  static void callMinimax(Integer depth, Boolean isMaximizer) {
    nodes.clear();
    minimax(depth, isMaximizer);
  }

  static void showBoard() {
    for (int i = 0; i < cells.length; i++) {
      for (int j = 0; j < cells.length; j++) {
        if (cells[i][j] == 0) {
          points[i][j] = '-';
        }
        if (cells[i][j] == 1) {
          points[i][j] = 'X';
        }
        if (cells[i][j] == -1) {
          points[i][j] = 'O';
        }
        System.out.printf("[ %s ]", points[i][j]);
      }
      System.out.println();
    }
  }

  static void showHint() {
    for (int i = 0; i < cells.length; i++) {
      for (int j =0; j < cells.length; j++) {
        System.out.printf("[ %d  %d ]", i + 1, j + 1);
      }
      System.out.println();
    }
  }

  static void startGame() {
    System.out.println("Welcome to the Tic-Tac-Toe game \n Author: Kingsley Victor \n");
    showBoard();
    showHint();
    System.out.println("\n To place a mark, enter the corresponding row and column value \n");
    Scanner gameInput = new Scanner(System.in);
    System.out.println("\n Who plays first?  Computer (1), User (2) \n");
    Integer choice = gameInput.nextInt();
    while (choice < 1 || choice > 2) {
      System.out.println("Invalid choice");
      choice = gameInput.nextInt();
    }
    if (choice == 1) {
      Random random = new Random();
      Move m = new Move(random.nextInt(3), random.nextInt(3));
      placeMove(m, 1);
      showBoard();
      showHint();
    }
    while (!gameIsOver()) {
      System.out.println("Make your move \n \n");
      Integer r = gameInput.nextInt() - 1;
      Integer c = gameInput.nextInt() - 1;
      Move m = new Move(r, c);
      while (cells[m.getR()][m.getC()] != 0) {
        System.out.println("Invalid move. Cell already occupied \n \n");
        r = gameInput.nextInt() - 1;
        c = gameInput.nextInt() - 1;
        m = new Move(r, c);
      }
      placeMove(m, -1);
      showBoard();
      callMinimax(0, true);
      if (gameIsOver()) {
        System.out.println();
        showBoard();
        break;
      }
      try {
        System.out.println();
        System.out.println("Wait! Computer is thinking");
        System.out.println();
        Thread.sleep(2000);
      } catch (InterruptedException e) {
        System.out.println(e.getMessage());
      }
      placeMove(getBestMove(), 1);
      showBoard();
      if (!gameIsOver()) {
        showHint();
      }
    }
    if (xHasWon()) {
      System.out.println("Computer wins the game");
    }
    if (oHasWon()) {
      System.out.println("You win the game");
    }
    if (getAvailableMoves().isEmpty() && !xHasWon() && !oHasWon()) {
      System.out.println("It's a tie");
    }
  }
}

public class Game {
  public static void main(String[] args) {
    Board.startGame();
  }
}

File: src/Game.js
const readLine = require('readline');

class Move {
  /**
   * 
   * @param {number} r 
   * @param {number} c 
   */
  constructor(r, c) {
    this.r = r;
    this.c = c;
  }

  getR() {
    return this.r;
  }

  getC() {
    return this.c;
  }
}



class Point {
  /**
   * 
   * @param {number} value 
   * @param {Move} move 
   */
  constructor(value, move) {
    this.value = value;
    this.move = move;
  }

  getValue() {
    return this.value;
  }

  getMove() {
    return this.move;
  }
}

class Board {
  constructor() {
    /**
     * @type number[][]
     */
    this.cells = [[0, 0, 0], [0, 0, 0], [0, 0, 0]];
    /**
     * @type Move[]
     */
    this.availableMoves = [];
    /**
     * @type Point[]
     */
    this.nodes = [];
    /**
     * @type string[][]
     */
    this.marks = [['-', '-', '-'], ['-', '-', '-'], ['-', '-', '-']];
  }

  /**
   * 
   * @param {Move} m 
   * @param {number} i 
   */
  placeMove(m, i) {
    this.cells[m.getR()][m.getC()] = i;
  }

  getAvailableMoves() {
    if (this.availableMoves.length >  0) {
      for (let i = 0; i < this.availableMoves.length; i++) {
        this.availableMoves.splice(i, 1);
      }
    }
    for (let i = 0; i < this.cells.length; i++) {
      for (let j =0; j < this.cells.length; j++) {
        if (this.cells[i][j] === 0) {
          this.availableMoves.push(new Move(i, j));
        }
      }
    }
    return this.availableMoves;
  }

  /**
   * 
   * @param {number} value 
   */
  thereIsAWin(value) {
    for (let i =0; i < this.cells.length; i++) {
      if (this.cells[i][0] === this.cells[i][1] && this.cells[i][1] === this.cells[i][2] && this.cells[i][2] === value) {
        return true;
      }
      if (this.cells[0][i] === this.cells[1][i] && this.cells[1][i] === this.cells[2][i] && this.cells[2][1] === value) {
        return true;
      }
    }
    if ((this.cells[0][0] === this.cells[1][1] && this.cells[1][1] === this.cells[2][2] && this.cells[2][2] === value) || (this.cells[0][2] === this.cells[1][1] && this.cells[1][1] === this.cells[2][0] && this.cells[2][0] === value)) {
      return true;
    }
    return false;
  }

  xHasWon() {
    return this.thereIsAWin(1);
  }

  oHasWon() {
    return this.thereIsAWin(-1);
  }

  gameIsOver() {
    return (this.xHasWon() || this.oHasWon() || this.getAvailableMoves().length === 0);
  }

  /**
   * 
   * @param {number} depth 
   * @param {boolean} isMaximizer 
   */
  minimax(depth, isMaximizer) {
    let maxEval = Number.MIN_VALUE;
    let minEval = Number.MAX_VALUE;
    if (this.getAvailableMoves().length === 0) {
      return 0;
    }
    if (this.xHasWon()) {
      return 1;
    }
    if (this.oHasWon()) {
      return -1;
    }
    if (isMaximizer) {
      for (let i = 0; i < this.getAvailableMoves().length; i++) {
        const m = this.getAvailableMoves()[i];
        this.placeMove(m, 1);
        const evaluate = this.minimax(depth + 1, false);
        maxEval = Math.max(maxEval, evaluate);
        if (depth === 0) {
          this.nodes.push(new Point(maxEval, m));
        }
        this.placeMove(m, 0);
      }
    } else {
      for (let i = 0; i < this.getAvailableMoves().length; i++) {
        const m = this.getAvailableMoves()[i];
        this.placeMove(m, -1);
        const evaluate = this.minimax(depth + 1, true);
        minEval = Math.min(minEval, evaluate);
        if (depth === 0) {
          this.nodes.push(new Point(minEval, m));
        }
        this.placeMove(m, 0);
      }
    }
    // console.log(this.minimax(depth, isMaximizer));
    return isMaximizer ? maxEval : minEval;
  }

  getBestMove() {
    let max = Number.MIN_VALUE;
    let index = 0;
    for (let i = 0; i < this.nodes.length; i++) {
      const node = this.nodes[i];
      if (max < node.getValue()) {
        max = node.getValue();
        index = i;
      }
    }
    return this.nodes[index].getMove();
  }

  /**
   * 
   * @param {number} depth 
   * @param {boolean} isMaximizer 
   */
  callMinimax(depth, isMaximizer) {
    for (let i = 0; i < this.nodes.length; i++) {
      this.nodes.splice(i, 1);
    }
    this.minimax(depth, isMaximizer);
  }

  showBoard() {
    for (let i = 0; i < this.cells.length; i++) {
      for (let j = 0; j < this.cells.length; j++) {
        if (this.cells[i][j] === 0) {
          this.marks[i][j] = '-';
        }
        if (this.cells[i][j] === 1) {
          this.marks[i][j] = 'X';
        }
        if (this.cells[i][j] === -1) {
          this.marks[i][j] = 'O';
        }
        process.stdout._write('[ '+ this.marks[i][j] + ' ]', 'utf-8', (err) => {});
      }
      console.log('\n');
    }
  }

  showHint() {
    for (let i = 0; i < this.cells.length; i++) {
      for (let j = 0; j < this.cells.length; j++) {
        process.stdout._write(`[${[ i + 1, j + 1 ]}]`, 'utf-8', (err) => {});
      }
      console.log('\n');
    } 
  }

  async startGame() {
    console.log('Welcome to the Tic-Tac-Toe game \n Author: Kingsley Victor \n');
    this.showBoard();
    this.showHint();
    console.log('\n To place a mark, enter the corresponding row and column value \n');
    const reading = readLine.createInterface({
      input: process.stdin,
      output: process.stdout
    });
    let getChoice = await new Promise((resolve) => {
      reading.question('\n Who plays first? Computer(1), User (2) \n', (data) => {
        resolve(data);
      });
    });
    const checkChoice = async () => {
      getChoice = await new Promise((resolve) => {
        reading.question('\n Invalid Choice \n', (data) => {
          resolve(data);
        });
      });
      if (parseInt(getChoice) < 1 || parseInt(getChoice) > 2) {
        checkChoice();
      }
    }
    if (parseInt(getChoice) < 1 || parseInt(getChoice) > 2) {
      checkChoice();
    }
    if (parseInt(getChoice) === 1) {
      const randomRow = Math.floor(Math.random() * 3);
      const randomColumn = Math.floor(Math.random() * 3);
      const m = new Move(randomRow, randomColumn);
      this.placeMove(m, 1);
      this.showBoard();
      this.showHint();
    }
    const checkIfGameOver = async () => {
      console.log('Make your move \n \n');
      let r = await new Promise((resolve) => {
        reading.question('\n Enter Row \n', (data) => {
          resolve(parseInt(data) - 1);
        });
      });
      let c = await new Promise((resolve) => {
        reading.question('\n Enter Column \n', (data) => {
          resolve(parseInt(data) - 1);
        });
      });
      let m = new Move(r, c);
      const checkIfCellOccupied = async () => {
        console.log('\n Invalid move. Cell already occupied \n');
        r = await new Promise((resolve) => {
          reading.question('\n Enter Row \n', (data) => {
            resolve(parseInt(data) - 1);
          });
        });
        c = await new Promise((resolve) => {
          reading.question('\n Enter Column \n', (data) => {
            resolve(parseInt(data) - 1);
          });
        });
        m = new Move(r, c);
        if (this.cells[m.getR()][m.getC()] !== 0) {
          checkIfCellOccupied();
        }
      }
      if (this.cells[m.getR()][m.getC()] !== 0) {
        checkIfCellOccupied();
      } else {
        this.placeMove(m, -1);
        this.showBoard();
        this.callMinimax(0, true);
        if (this.gameIsOver()) {
        console.log('\n');
        this.showBoard();
        reading.close();
      }
      try {
        console.log();
        console.log('Wait! Computer is thinking');
        console.log()
        setTimeout(() => {}, 2000);
      } catch (error) {
        console.log(error.message);
      }
      this.placeMove(this.getBestMove(), 1);
      this.showBoard();
      if (!this.gameIsOver()) {
        this.showHint();
        checkIfGameOver();
      }
      }
    }
    if (!this.gameIsOver()) {
      checkIfGameOver();
    }
    if (this.xHasWon()) {
      console.log('Computer wins the game');
    }
    if (this.oHasWon()) {
      console.log('You win the game');
    }
    if (this.getAvailableMoves().length === 0 && !this.xHasWon() && !this.oHasWon()) {
      console.log(`It's a tie`);
    }
  }
}

class Game {
  start() {
    const b = new Board();
    b.startGame();
  }
}

const game = new Game();

game.start();
